File: src/tco.py
import logging
import re
from typing import Dict, Tuple

logger = logging.getLogger(__name__)


def _parse_capacity_to_gb(capacity_str: str) -> int:
    """Convert strings like '8GB', '1TB' to integer GB. Returns 0 when unknown."""
    if not capacity_str or capacity_str.upper() == 'N/A':
        return 0
    capacity_str = capacity_str.upper()
    match = re.search(r"(\d+\.\d+|\d+)", capacity_str)
    if not match:
        return 0
    val = float(match.group(1))
    if 'TB' in capacity_str:
        val *= 1024
    return int(val)


def calculate_tco_and_perf(item: Dict, assumptions: Dict) -> Tuple[float, float]:
    """Return (tco, performance_per_dollar) following the same rules as frontend JS.

    item: enriched listing dict from enrich_item (contains price, cpu_idle_power, etc.)
    assumptions: dict from config['app']['tco_assumptions']
    """
    if item.get('cpu_idle_power') in (None, '') or item.get('price') in (None, ''):
        return None, None

    price = float(item['price'])
    idle_watts = float(item['cpu_idle_power'])

    kwh_cost = float(assumptions.get('kwh_cost', 0.14))
    lifespan_years = max(int(assumptions.get('lifespan_years', 5)), 1)

    energy_cost = (idle_watts / 1000) * 24 * 365 * lifespan_years * kwh_cost

    # shipping
    shipping_cost = 0.0
    if not item.get('free_shipping'):
        if str(item.get('cpu_model', '')).upper().endswith('T'):
            shipping_cost = float(assumptions.get('shipping_cost_t_cpu', 10))
        else:
            shipping_cost = float(assumptions.get('shipping_cost_non_t_cpu', 35))

    # RAM shortfall
    item_ram_gb = _parse_capacity_to_gb(item.get('ram'))
    required_ram = float(assumptions.get('required_ram_gb', 16))
    ram_shortfall_cost = float(assumptions.get('ram_upgrade_flat_cost', 30)) if item_ram_gb < required_ram else 0.0

    # Storage shortfall
    item_storage_gb = _parse_capacity_to_gb(item.get('storage'))
    required_storage = float(assumptions.get('required_storage_gb', 128))
    storage_shortfall_cost = float(assumptions.get('storage_upgrade_flat_cost', 15)) if item_storage_gb < required_storage else 0.0

    # AC adapter assumed included for automated job
    ac_adapter_cost = 0.0

    tco = price + energy_cost + shipping_cost + ram_shortfall_cost + storage_shortfall_cost + ac_adapter_cost

    perf = item.get('performance')
    perf_per_dollar = (float(perf) / tco) if perf and tco > 0 else None

    return tco, perf_per_dollar 

File: static/js/tco.js
// TCO related functions will be moved here

function calculateTCOJS(item, assumptions, overrides = {}) {
    const itemCPU = item.cpu_model || "unknown";
    const itemId = item.itemId;

    console.log(`[TCO_CALC_START] Item: ${itemCPU} (${itemId}) | Price: ${item.price} | Idle Power: ${item.cpu_idle_power} | Overrides: ${JSON.stringify(overrides)} | Assumptions: ${JSON.stringify(assumptions)}`);

    if (item.cpu_idle_power === null || item.cpu_idle_power === undefined) {
        console.log(`[TCO_CALC_END] Item: ${itemCPU} (${itemId}) - Result: N/A (No idle power data)`);
        return { tco: null, performance_per_dollar: null };
    }

    if (item.price === null || item.price === undefined) {
        console.log(`[TCO_CALC_END] Item: ${itemCPU} (${itemId}) - Result: N/A (No price data)`);
        return { tco: null, performance_per_dollar: null };
    }

    const numericPrice = parseFloat(item.price);
    const idleWatts = parseFloat(item.cpu_idle_power);
    
    const kwhCost = parseFloat(assumptions.kwh_cost);
    const lifespanYears = parseInt(assumptions.lifespan_years, 10);
    // Ensure lifespanYears is at least 1 for calculation, but log the original if it was <= 0
    let safeLifespanYears = lifespanYears;
    if (lifespanYears <= 0) {
        console.log(`[TCO_INFO] Item: ${itemCPU} (${itemId}) - Lifespan years was ${lifespanYears}, using 1 for energy calculation.`);
        safeLifespanYears = 1;
    }

    const energyCostCalc = (idleWatts / 1000) * (24 * 365 * safeLifespanYears) * kwhCost;
    
    let shippingCostToAdd = 0;
    const defaultShippingT = parseFloat(assumptions.shipping_cost_t_cpu);
    const defaultShippingNonT = parseFloat(assumptions.shipping_cost_non_t_cpu);

    if (overrides.shipping !== undefined && overrides.shipping !== null && overrides.shipping !== '') {
        shippingCostToAdd = parseFloat(overrides.shipping);
    } else if (!item.free_shipping) {
        if (item.cpu_model && typeof item.cpu_model === 'string' && item.cpu_model.toUpperCase().endsWith('T')) {
            shippingCostToAdd = defaultShippingT;
        } else {
            shippingCostToAdd = defaultShippingNonT;
        }
    }
    
    let ramShortfallCost = 0;
    const itemRamGb = parseCapacityToGBJS(item.ram);
    const requiredRamGb = parseFloat(assumptions.required_ram_gb);
    const ramUpgradeCost = parseFloat(assumptions.ram_upgrade_flat_cost);
    if (itemRamGb < requiredRamGb) {
        ramShortfallCost = ramUpgradeCost;
    }
    
    let storageShortfallCost = 0;
    const itemStorageGb = parseCapacityToGBJS(item.storage);
    const requiredStorageGb = parseFloat(assumptions.required_storage_gb);
    const storageUpgradeCost = parseFloat(assumptions.storage_upgrade_flat_cost);
    if (itemStorageGb < requiredStorageGb) {
        storageShortfallCost = storageUpgradeCost;
    }

    let acAdapterCost = 0;
    const acAdapterActuallyIncluded = overrides.ac_adapter_included === undefined ? true : overrides.ac_adapter_included;
    if (!acAdapterActuallyIncluded) {
        acAdapterCost = 10; // Hardcoded $10 for AC adapter
    }
    
    const tco = numericPrice + energyCostCalc + shippingCostToAdd + ramShortfallCost + storageShortfallCost + acAdapterCost;
    
    let performancePerDollar = null;
    if (item.performance && tco && tco > 0) {
        performancePerDollar = parseFloat(item.performance) / tco;
    }

    console.log(`[TCO_BREAKDOWN] Item: ${itemCPU} (${itemId})`);
    console.log(`  Price: ${numericPrice.toFixed(2)}`);
    console.log(`  Energy Cost: ${energyCostCalc.toFixed(2)} (Idle: ${idleWatts}W, Lifespan: ${safeLifespanYears}yrs, kWh Cost: ${kwhCost})`);
    console.log(`  Shipping Cost: ${shippingCostToAdd.toFixed(2)} (Free: ${item.free_shipping}, Override: ${overrides.shipping}, Default T: ${defaultShippingT}, Default Non-T: ${defaultShippingNonT})`);
    console.log(`  RAM Shortfall: ${ramShortfallCost.toFixed(2)} (Item RAM: ${itemRamGb}GB, Required: ${requiredRamGb}GB, Upgrade Cost: ${ramUpgradeCost})`);
    console.log(`  Storage Shortfall: ${storageShortfallCost.toFixed(2)} (Item Storage: ${itemStorageGb}GB, Required: ${requiredStorageGb}GB, Upgrade Cost: ${storageUpgradeCost})`);
    console.log(`  AC Adapter Cost: ${acAdapterCost.toFixed(2)} (Included by override/default: ${acAdapterActuallyIncluded})`);
    console.log(`  --------------------`);
    console.log(`  Calculated TCO: ${isNaN(tco) ? 'N/A' : tco.toFixed(2)}`);
    console.log(`  Performance Score: ${item.performance || 'N/A'}`);
    console.log(`  Calculated Perf/$: ${performancePerDollar ? (isNaN(performancePerDollar) ? 'N/A' : performancePerDollar.toFixed(1)) : 'N/A'}`);
    console.log(`[TCO_CALC_END] Item: ${itemCPU} (${itemId}) - Assumptions Used: ${JSON.stringify(assumptions)}, Overrides Used: ${JSON.stringify(overrides)}`);
    
    return { tco: isNaN(tco) ? null : tco, performance_per_dollar: (performancePerDollar && !isNaN(performancePerDollar)) ? performancePerDollar : null };
}

function populateTCOFields(tcoDefaults) {
    if (!tcoDefaults) return;
    document.getElementById('kwh_cost').value = tcoDefaults.kwh_cost;
    document.getElementById('lifespan_years').value = tcoDefaults.lifespan_years;
    document.getElementById('shipping_cost_t_cpu').value = tcoDefaults.shipping_cost_t_cpu;
    document.getElementById('shipping_cost_non_t_cpu').value = tcoDefaults.shipping_cost_non_t_cpu;
    document.getElementById('required_ram_gb').value = tcoDefaults.required_ram_gb;
    document.getElementById('ram_upgrade_flat_cost').value = tcoDefaults.ram_upgrade_flat_cost;
    document.getElementById('required_storage_gb').value = tcoDefaults.required_storage_gb;
    document.getElementById('storage_upgrade_flat_cost').value = tcoDefaults.storage_upgrade_flat_cost;
}

function toggleTCOAssumptions() {
    const form = document.getElementById('tcoAssumptionsForm');
    const button = document.getElementById('tcoToggleButton');
    if (form.style.display === 'none') {
        form.style.display = 'grid'; 
        button.textContent = 'Adjust TCO Assumptions ▲';
    } else {
        form.style.display = 'none';
        button.textContent = 'Adjust TCO Assumptions ▼';
    }
} 